### ISSUE TITLE
support for yara "private rule"

### ISSUE THREAD

**Issue by securiteinfo** [Cisco-Talos/clamav#1153]
Hi,

private rules in yara allow to have C like functions. It works with yara, but clamav does not support it.

Example : These are test rules : 
cat test.yara
private rule Rule1
{
    strings:
        $a = "rulez" private

    condition:
        $a
}

private rule Rule2
{
    strings:
        $a = "www.toto.com" private

    condition:
        $a
}

rule MainRule
{
    condition:
        any of (Rule*)
}

running yara : 

yara test.yara .
MainRule ./test.yara

running clamscan : 

clamscan -d test.yara
LibClamAV Error: yyerror(): test.yara line 4 syntax error, unexpected _PRIVATE_, expecting _CONDITION_
LibClamAV Error: yyerror(): test.yara line 13 syntax error, unexpected _PRIVATE_, expecting _CONDITION_
LibClamAV Error: yyerror(): test.yara line 22 syntax error, unexpected _IDENTIFIER_, expecting _STRING_IDENTIFIER_ or _STRING_IDENTIFIER_WITH_WILDCARD_
LibClamAV Warning: cli_loadyara: failed to parse or load 3 yara rules from file test.yara, successfully loaded 0 rules.
LibClamAV Warning: cli_loadyara: empty database file

**Comment by micahsnyder** [comment]
I wouldn't call it "C-like", but yes very powerful.

We plan to add a very similar feature to clamav.  It's something we've been calling "weak indicators".  They would be clamav signatures that do not alert, but that you could place in the logical condition.  Same idea.  Except we want to take it a step further and let you say that the weak indicator / private rule may have matched on a child-layer.  For example you could have a signature that matches on content in an email, and have the condition also depend on a weak indicator match on an email attachment. 

So in short yes this is a basically planned feature since we also have plans to focus on yara support in the future, and one that we wish to take even further than how Yara does it.

**Comment by HydraDragonAntivirus** [comment]
I have an idea to how to create private rule first create rule as public in ClamAV then create privaterule.ign2 then exclude private rule.
